### PR TITLE
Add Moment.to_text_diagram

### PR DESCRIPTION
**Pull request by Strilanc**
- Creates a 2d view of the moment
- `Moment.__str__` now produces a text diagram
- Moved several diagram utilities out of circuit and into private methods inside the diagram info protocol file
- Added `__complex__`/`__float__`/`__int__` to GridQubit/LineQubit as appropriate

This was motivated by being confused by a large circuit on a grid, and wanting to see some of the 2d structure more directly.

Example diagram:

```
import cirq
a, b, c, d = cirq.GridQubit.rect(2, 2)
print(cirq.Moment(cirq.S(c), cirq.ISWAP(a, d))
```

```
  ╷ 0     1
╶─┼─────────────
0 │ iSwap─┐
  │       │
1 │ S     iSwap
  │
```

### CODE REVIEW

**Review by maffoo — APPROVED** [review]
Minor comments, then LGTM

File: cirq/circuits/circuit.py [hunk 1]
```diff
@@ -2053,22 +1972,13 @@ def _draw_moment_in_diagram(
             out_diagram.vertical_line(x, y1, y2)
 
         # Print gate qubit labels.
-        for s, q in zip(info.wire_symbols, op.qubits):
+        symbols = info._wire_symbols_including_formatted_exponent(
+            args,
+            preferred_exponent_index=max([i for i in range(len(op.qubits))],
```
[the remark this answers is not in the document]
Comment (Strilanc) [reply cirq/circuits/circuit.py]: Done.

File: cirq/devices/grid_qubit.py [hunk 2]
```diff
@@ -72,6 +72,9 @@ def neighbors(self, qids: Optional[Iterable[ops.Qid]] = None
     def _with_row_col(self: TSelf, row: int, col: int) -> TSelf:
         """Returns a qid with the same type but a different coordinate."""
 
+    def __complex__(self):
+        return self.col + 1j * self.row
```
[the remark this answers is not in the document]
Comment (Strilanc) [reply cirq/devices/grid_qubit.py:76]: Done.

File: cirq/devices/line_qubit.py [hunk 3]
```diff
@@ -99,6 +99,15 @@ def __rsub__(self: TSelf, other: int) -> TSelf:
...
         return self._with_x(-self.x)
 
+    def __complex__(self):
+        return complex(self.x)
+
+    def __float__(self):
+        return float(self.x)
+
+    def __int__(self):
+        return int(self.x)
```
[the remark this answers is not in the document]
Comment (Strilanc) [reply cirq/devices/line_qubit.py:109]: Done.

File: cirq/ops/moment.py [hunk 4]
```diff
@@ -308,3 +317,108 @@ def __getitem__(self, key):
...
+        """
+
+        # Figure out where to place everything.
+        qs = set(self.qubits) | set(extra_qubits)
+        points = {xy_breakdown_func(q) for q in qs}
+        x_keys = sorted({pt[0] for pt in points}, key=_SortByValFallbackToType)
+        y_keys = sorted({pt[1] for pt in points}, key=_SortByValFallbackToType)
+        x_map = {x_key: x + 2 for x, x_key in enumerate(x_keys)}
+        y_map = {y_key: y + 2 for y, y_key in enumerate(y_keys)}
+        qubit_map = {}
```
[the remark this answers is not in the document]
Comment (Strilanc) [reply cirq/ops/moment.py]: Done.

**Pull request merged**